calcom/cal.com
# feat: Use team logos for various meta icons

## What does this PR do?

We're changing all of our meta icons (favicons, apple touch icons, android chrome icons, mstile icons) over to use `api/logo` to ensure the correct images are displayed for teams that are using custom branding.

For this first draft, I am resolving all of the different types of icons to the existing `Team.appLogo` or `Team.appIconLogo`. There is currently no UI for setting those logos, so I don't know what expectations – if any – we have for the size requirements of either.

(Update: 2023-05-18) Our various meta icons expect to be served in varying sizes. With this PR, the logo endpoint now uses the Next.js built-in image processing to resize team logos down to appropriate sizes.

The image processing is only invoked on custom team icons, and only for icons that have specific size requirements (all the meta icons).

Requests for the type "logo" or "icon" are passed through without processing. If we ever fall back to Cal.com default icons (non-team pages, or teams without custom logos) then the fallback icons are also served without any additional processing.



Fixes #8441 (issue)



**Environment**: Staging(main branch) / Production

## Type of change



- Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

We must test both the team-specific and fallback behaviour of this change, for each type of icon.

To test original, or fallback, behaviour we can just use the regular development environment on localhost:3000.

To test team-specific logos, I had:

1. Created a team `cool-team` via Cal.com
2. Edited the team in the database, copying the data from its `logo` column into `appLogo` and `appIconLogo`
3. Set my machine up to resolve `cool-team.example.test` to 127.0.0.1
4. And then I'm able to access the team-branded version of the site via `cool-team.example.test:3000`.

When visiting each domain we see either custom or regular favicons:

Scenario | Screenshot
---- | ---
Team with custom logos | ![image](https://github.com/calcom/cal.com/assets/602850/0dac6ac8-4a71-4d09-8758-28d790eb888c)
Team without logos | ![image](https://github.com/calcom/cal.com/assets/602850/53c47163-18bf-4a81-9bd1-b32389829f64)
Non-team domain | ![image](https://github.com/calcom/cal.com/assets/602850/51d31cca-e392-45b7-abf6-ddde14810134)
Localhost domain | ![image](https://github.com/calcom/cal.com/assets/602850/f023cb1b-1fe0-4aae-a2a1-026072a5b303)

And finally, the api/logo endpoint resizes images appropriately for the different types of image supported:

Type | Team output | Default output |
--- | --- | ---
unspecified | ![cool-team-1-unspecified](https://github.com/calcom/cal.com/assets/602850/a244b89a-ce17-4bec-9092-845dd8ae5068) | ![regular-team-1-unspecified](https://github.com/calcom/cal.com/assets/602850/be2b0c5a-e47b-42df-bb90-e94371954581)
logo | ![cool-team-2-logo](https://github.com/calcom/cal.com/assets/602850/5f387273-b521-43f2-b676-0705975e5a73) | ![regular-team-2-logo](https://github.com/calcom/cal.com/assets/602850/756490be-2ac1-4d5c-844a-a8c826e61a22)
icon | ![cool-team-3-icon](https://github.com/calcom/cal.com/assets/602850/af3124fc-a0bb-49a5-abae-e551f2039b0d) | ![regular-team-3-icon](https://github.com/calcom/cal.com/assets/602850/ef17b2af-8443-4374-980b-7dff83b91bf2)
favicon-16 | ![cool-team-4-favicon-16](https://github.com/calcom/cal.com/assets/602850/98a0e09e-394d-434b-9d90-e2a5ce91d46a) | ![regular-team-4-favicon-16](https://github.com/calcom/cal.com/assets/602850/db732eaa-2d53-47ab-a23a-012d4eee50de)
favicon-32 | ![cool-team-5-favicon-32](https://github.com/calcom/cal.com/assets/602850/52216248-0cfd-4cde-9d31-3cd38768e259) | ![regular-team-5-favicon-32](https://github.com/calcom/cal.com/assets/602850/a887bfda-e4c8-48e1-9f9f-dff07c65f757)
apple-touch-icon | ![cool-team-6-apple-touch-icon](https://github.com/calcom/cal.com/assets/602850/165f69cb-c54b-4a21-a571-9870b9df0aae) | ![regular-team-6-apple-touch-icon](https://github.com/calcom/cal.com/assets/602850/25a2a103-50b0-4557-b291-354871e3c148)
mstile | ![cool-team-7-mstile](https://github.com/calcom/cal.com/assets/602850/e2384db4-6e32-4958-b101-5c818c859d96) | ![regular-team-7-mstile](https://github.com/calcom/cal.com/assets/602850/92b117b4-31c2-4599-8e32-fbbe5ecc12b3)
android-chrome-192 | ![cool-team-8-android-chrome-192](https://github.com/calcom/cal.com/assets/602850/10170d21-b85a-49ce-940c-bea87214d547) | ![regular-team-8-android-chrome-192](https://github.com/calcom/cal.com/assets/602850/0fe4f9fb-b7fc-4a20-bee9-e50200129373)
android-chrome-256 | ![cool-team-9-android-chrome-256](https://github.com/calcom/cal.com/assets/602850/244f01cc-dbf9-484f-bcd5-f3e81f5a6577) | ![regular-team-9-android-chrome-256](https://github.com/calcom/cal.com/assets/602850/7800aa78-5dd4-4279-9a0b-3d0aa141c6e6)

## Checklist



- I haven't added tests that prove my fix is effective or that my feature works

/claim #8441
